### PR TITLE
error checking that kl_div's input <= 0

### PR DESCRIPTION
**Pull request by mengban**
When I used torch.nn.KLDivLoss(x,y) as my loss function of my model.However,the negative loss value was calculated.In fact,the y should be log_softmax.
So I added an assertion to fix the bug.

### CODE REVIEW

**Review by ssnl — CHANGES_REQUESTED** [review]
This synchronizes on CUDA.... if you want to add the check instead move it to the kernel code.

**Comment by mengban** [comment]
> This synchronizes on CUDA.... if you want to add the check instead move it to the kernel code.
Do you mean i should add the check in 
[https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/loss.py#L296]()?

**Comment by SsnL** [comment]
No, this would involve rewriting `kl_div` for GPU as one ore more CUDA kernels and add a device assert in one of them: 
https://github.com/pytorch/pytorch/blob/65b00aa5972e23b2a70aa60dec5125671a3d7153/aten/src/ATen/native/Loss.cpp#L63

**Comment by mengban** [comment]
> No, this would involve rewriting `kl_div` for GPU as one or more CUDA kernels and add a device assert in one of them:
> [pytorch/aten/src/ATen/native/Loss.cpp](https://github.com/pytorch/pytorch/blob/65b00aa5972e23b2a70aa60dec5125671a3d7153/aten/src/ATen/native/Loss.cpp#L63)
> 
> Line 63 in [65b00aa](/pytorch/pytorch/commit/65b00aa5972e23b2a70aa60dec5125671a3d7153)
> 
>  Tensor kl_div(const Tensor& input, const Tensor& target, int64_t reduction) { 
> That is a involved change though..

emmmm... so where should i add the check

**Comment by fmassa** [comment]
@mengban this seems to be an involved task just to add the check in an efficient manner. It would involve rewriting `kl_divergence` in CUDA (probably using TensorIterator), and is a significant amount of work.

That being said, the [documentation explicitly states](https://pytorch.org/docs/stable/nn.html?highlight=kl_div#torch.nn.KLDivLoss) that the inputs should be log-probabilities.

**Comment by mengban** [comment]
> @mengban this seems to be an involved task just to add the check in an efficient manner. It would involve rewriting `kl_divergence` in CUDA (probably using TensorIterator), and is a significant amount of work.
> 
> That being said, the [documentation explicitly states](https://pytorch.org/docs/stable/nn.html?highlight=kl_div#torch.nn.KLDivLoss) that the inputs should be log-probabilities.

Thanks a lot. But why not implement inside the log-probabilities function?

**Comment by SsnL** [comment]
@mengban because there is no function specifically used for log of probabilities. The kl loss calls regular `log`.

**Comment by mengban** [comment]
@SsnL Got it. Thank u~
